core/src/toga/images.py

```diff
@@ -99,6 +99,9 @@ def __init__(
             src.save(buffer, format="png", compress_level=0)
             self._impl = self.factory.Image(interface=self, data=buffer.getvalue())
 
+        elif isinstance(src, self.factory.Image.RAW_TYPE):
```

Comment:
@rmartin16 If you're looking for a hard-mode typing challenge: How should the type annotation on `src` be modified to allow for *this*? It's a type on a class that, by definition, is platform specific, and won't exist in the environment where toga-core is being built, tested, or documented. 

The only solution I can think of is to add `UIImage | NSImage | GdkBitmap | ...` but that won't be correct - you *can't* use a UIImage *unless* you're on iOS. 

Any ideas?

Reply:
I think this might be possible....but it, perhaps, starts to verge on the ridiculous....

For instance, here's where I got to before I, more or less, gave up:
```python
if TYPE_CHECKING:
    BaseImageSrcT: TypeAlias = Union[
        Path,
        bytes,
        bytearray,
        memoryview,
        "Image",
        PIL.Image.Image,
        None,
    ]

    if sys.platform == "win32":
        from toga_winforms.images import WinImage
        ImageSrcT: TypeAlias = Union[BaseImageT, toga_winforms.images.Image]

    if sys.platform.startswith("linux"):
        from toga_gtk.images import GdkPixbuf
        ImageSrcT: TypeAlias = Union[BaseImageT, GdkPixbuf.Pixbuf]

    if sys.platform == "darwin":
        from toga_cocoa.images import NSImage
        ImageSrcT: TypeAlias = Union[BaseImageT, toga_gtk.images.Image]


class Image:
    def __init__(self, src: ImageSrcT): ...
```

I'm not sure if mypy technically considers this [strategy above](https://mypy.readthedocs.io/en/stable/common_issues.html#version-and-platform-checks) part of their larger [type narrowing](https://mypy.readthedocs.io/en/stable/type_narrowing.html) support, but it helps accommodate the runtime environment.

That said, for this to really make mypy happy, it'll require other changes in Toga...notably, replacing the star imports in the platform-specific Toga code to explicitly import these native library classes....and we'll also need typing stubs for, at least, PyGObject.

BUT...this has major consequences for building docs. On my Linux machine, it tries to interpret the definition of `ImageSrcT` for Linux but cannot find `toga_gtk`; I can install `toga_gtk` to build the docs....but then the build blows up with `RuntimeError`s about `DISPLAY` env var. Maybe Sphinx has some buried support for effectively ignoring certain symbols...and that could serve as some type of resolution here.

At any rate, the `src` argument feels like a good candidate for `Any` where the documentation fills in the details.

P.S. - type checkers and IDEs are still likely to resolve the type of `src` to `Any` even with all these accommodations if Pillow isn't installed. And if we just add these platform-specific image classes, such as `GdkPixbuf.Pixbuf`, then the type will _always_ be `Any` because it would be impossible to resolve all the types in the `Union` on any one system.

Reply:
> I think this might be possible....but it, perhaps, starts to verge on the ridiculous....
> 
> For instance, here's where I got to before I, more or less, gave up:

That was broadly my initial thought as well - except that it falls down on iOS/Android (which will never be installed in the local environment where sys.platform will find it), and on custom backends, and in the Sphinx builds you mention later.

> That said, for this to really make mypy happy, it'll require other changes in Toga...notably, replacing the star imports in the platform-specific Toga code to explicitly import these native library classes....and we'll also need typing stubs for, at least, PyGObject.

Woof... good luck with *that* endeavour... I like yaks as much as the next person, but autogeneration of types from GI bindings is a yak to far for me :-)

> At any rate, the `src` argument feels like a good candidate for `Any` where the documentation fills in the details.

Is there any advantage to specifying the type *at all* in this case? What (if any) benefit exists to an explicit `Any` over "no type declaration"? Cosmetically, "Any" reads to me as "You can use literally anything", which isn't true (at least, not in spirit). This is a case where I think I'd prefer to say nothing, and force the reader to actually read words, rather than provide something explicitly too broad as an automated description.

Reply:
I'm not sure if this is actually better....but this is how the docs will render if a `NewType` is used instead of an implicit `TypeAlias`; while this still reveals the underlying type is `Any`, it's at least more forthright about something more being implied about this type....at least maybe anyway:
![image](https://github.com/beeware/toga/assets/387848/ad5b06fe-25ef-4960-99a8-b2f31a796d85)

Reply:
Although, something I'd need to test is whether using a `NewType` still actually allows `Any` without the caller needing the wrap the input with `ImageType()` to convince a type checker they are using this "new type"...

Reply:
Annoying that TypeAlias doesn't... you know... *alias* anything (at least, at the Sphinx level). 

However - I found [this post](https://github.com/pandas-dev/pandas/issues/33025#issuecomment-699636759) that gave a possible fix:
```
class ImageTypeClass:
    pass
ImageType = ImageTypeClass()
_ImageType: TypeAlias = Any
ImageType.__supertype__ = _ImageType
```

With that declaration, Sphinx renders: 
<img width="834" alt="Screenshot 2023-12-08 at 3 37 50 pm" src="https://github.com/beeware/toga/assets/37345/d492b587-81e5-410b-af01-aa8630780212">

which loses the hyperlinks, but gets the type in a helpful form. I have a suspicion that the hyperlink issue might be because ImageType doesn't have an autodoc declaration... but that doesn't explain why the hyperlinks for str etc are lost.

What does PyCharm make of that declaration?

Reply:
I believe what's happening here is Sphinx cannot determine what `ImageType` actually is...and when that happens, Sphinx seems to disable all typing for the entire function signature. This is a lot of what I dealt with while trying to get `TypeAlias` working for handlers.

As for PyCharm, it sees the type of `src` as `Any`. However, interestingly, when I just add `Any` to the `Union`, PyCharm shows the list of the other accepted types along with just accepting anything (such as a method):
![image](https://github.com/beeware/toga/assets/387848/06a274e3-a21e-4173-a8aa-6c84987e162d)

Reply:
Ok - here's a simpler version:
```
class ImageType:
    """Any valid image type, etc

    Not a real class. Don't try to create it.
    """

    __supertype__ = Any
```

which:
* renders Imagetype correctly;
<img width="830" alt="Screenshot 2023-12-09 at 8 15 34 am" src="https://github.com/beeware/toga/assets/37345/4934186a-e2e8-4903-9ac9-cf25f61e3c1c">
* provides a convenient place to document what an ImageType is
<img width="409" alt="Screenshot 2023-12-09 at 8 16 10 am" src="https://github.com/beeware/toga/assets/37345/b9ebdb03-3794-41bc-88f5-1e3d0a28fac6">

What does PyCharm do with that?

Reply:
(This is based on trying to work out what `__supertype__` actually does... AFAICT, it's the functional part of NewType)

Reply:
the cleverness will continue until the typing improves 😉 

Unfortunately, PyCharm does not seem to handle `__supertype__ = Any` as we might hope:
![image](https://github.com/beeware/toga/assets/387848/dd10667c-332c-4bb7-81eb-64ab4030051d)

Pylance, via VS Code doesn't seem to be a fan either...:
![image](https://github.com/beeware/toga/assets/387848/229d62ad-fa90-4095-b3f7-626f51ce6da7)

To top it off, though, `mypy` doesn't support this trick either:
```
src/togatypes/app.py:57: error: Argument "src" to "Image" has incompatible type "Image"; expected "Union[str, Path, bytes, bytearray, memoryview, ImageType, None]"  [arg-type]
src/togatypes/app.py:58: error: Argument "src" to "Image" has incompatible type "Image"; expected "Union[str, Path, bytes, bytearray, memoryview, ImageType, None]"  [arg-type]
```

Reply:
Additionally, fwiw, using `NewType` wouldn't really work either....since it would, in fact, require users to explicitly cast the image data as `ImageType` which wouldn't be a great experience at all.

Reply:
Ugh... *fine*, I guess we'll just go with `Any` in the union then... *le sigh*...

Reply:
The docs still look quite cluttered with everything in the union. Can we just make ImageType an alias for Any, and then describe all the options in its docstring?

Reply:
See the discussion above - that's where I started, and it looks awful; it renders in Sphinx as `NewType("Image", Any)`, with the sphinx link to "NewType" being to the Python docs. AFAIK, it's not possible to provide documentation for the type itself (which would be the ideal situation).

I'm open to other suggestions, but putting `Any` in the union is the only version I could find that mollifies mypy/PyCharm *and* doesn't read like a train wreck.

Reply:
Since a union that contains Any no longer provides any type-checking safety, and there are so many things in the union that it isn't very good documentation either, my suggestion was to just replace the whole union with Any, and rely on the documentation text to explain what's accepted.

Ideally we would do this with an alias, but if Sphinx doesn't work with that, we can use Any directly, and put the documentation in its own section on the Image page with incoming links from all the relevant places.